Force DWARF generation on windows with --dwarf-version

nurmukhametov opened the pull request:
This change lets `--dwarf-version=<n>` to force DWARF format debug info generation on Windows. It may be useful when ISPC code is linked with MINGW generated code. In that case, usual DWARF debug section (_debug_info, _debug_line, and etc.) are generated in PE files. This information is used by MINGW GDB during debug on Windows.

It fixes/implements #2129

e3m3 commented on src/ispc.h:
```diff
@@ -777,13 +777,16 @@ struct Globals {
     // xcrun dwarfdump -r0 object.o
     int generateDWARFVersion;
 
+    /** Default DWARF version to be used. */
+    int defaultDWARFVersion;
```
I think this default version should be implemented as a `constexpr`. Is there a use case where the default will change value?

dbabokin reviewed (commented):
The logic of the fix is really confusing. I failed to understand it.

We are going to have 3 different types of debug info (now we have 2, though) - `DWARF`, `CodeView` and `PDB` (not yet implemented), right? Should we have a variable `debugInfoType`, which would simplify and streamline the logic?

Also would be good to have lit tests for this.

nurmukhametov commented:
> The logic of the fix is really confusing. I failed to understand it.
> 
> We are going to have 3 different types of debug info (now we have 2, though) - `DWARF`, `CodeView` and `PDB` (not yet implemented), right? 

No, only `DWARF` and `CodeView`. I don't think that on `LLVM-IR` level there is `PDB` debug info type. Moreover, you can generate `PDB` file for project with `ISPC` code even now.

nurmukhametov commented:
>  Should we have a variable `debugInfoType`, which would simplify and streamline the logic?
> 
> Also would be good to have lit tests for this.

I have introduced `debugInfoType` variable and added lit tests.

aneshlya commented on tests/lit-tests/debug_type_dwarf.ispc:
```diff
@@ -0,0 +1,12 @@
+// RUN: %{ispc} -g --target=host --nostdlib --nowrap --emit-llvm-text -o - %s 2>&1 | FileCheck %s -check-prefix=CHECK_CODEVIEW
+// RUN: %{ispc} --dwarf-version=3 --target-os=windows --target=host --nostdlib --nowrap --emit-llvm-text -o - %s 2>&1 | FileCheck %s -check-prefix=CHECK_DWARF
+// RUN: %{ispc} -g --dwarf-version=3 --target-os=windows --target=host --nostdlib --nowrap --emit-llvm-text -o - %s 2>&1 | FileCheck %s -check-prefix=CHECK_DWARF
+// RUN: %{ispc} --dwarf-version=3 -g --target-os=windows --target=host --nostdlib --nowrap --emit-llvm-text -o - %s 2>&1 | FileCheck %s -check-prefix=CHECK_DWARF
```
The checks on line 5 and 6 are the same with different order of `-g`/`-dwarf-version` options. Is it intended?

nurmukhametov replied:
Yes, it is intended to check that `-g` logic doesn't redefine debug info type.

aneshlya commented on tests/lit-tests/debug_type_dwarf.ispc:
```diff
@@ -0,0 +1,12 @@
+// RUN: %{ispc} --target-os=windows --target=host --nostdlib --nowrap --emit-llvm-text -o - %s 2>&1 | FileCheck %s --implicit-check-not "llvm.dbg.cu"
+// RUN: %{ispc} -g --target-os=windows --target=host --nostdlib --nowrap --emit-llvm-text -o - %s 2>&1 | FileCheck %s -check-prefix=CHECK_CODEVIEW
+// RUN: %{ispc} -g --target=host --nostdlib --nowrap --emit-llvm-text -o - %s 2>&1 | FileCheck %s -check-prefix=CHECK_CODEVIEW
```
The test is supposed to be run on Windows host only, right? Please use `WINDOWS_HOST` in REQUIRES labels.

nurmukhametov replied:
The test is supposed to be run on any host that supports windows as target-os. I thought that `WINDOWS_ENABLED` means exactly that.

nurmukhametov replied:
Oh, yeah. I have fixed that.

aneshlya replied:
Now line 2 and 3 are the same :) please remove one of them

dbabokin commented on src/main.cpp:
```diff
@@ -782,6 +782,7 @@ int main(int Argc, char *Argv[]) {
             if (2 <= val && val <= 4) {
```
The latest DWARF version is 5. I'm not sure if LLVM supports it (and what version of LLVM supports it). I suggest checking it and if it's supported changing the range of supported versions.

nurmukhametov replied:
It looks like LLVM started to somehow support DWARF v5 since 7 ([link](https://releases.llvm.org/7.0.0/docs/ReleaseNotes.html)). I have added 5 as option as the separate commit.

dbabokin reviewed (approved):
Looks good to me, just minor remarks below.

nurmukhametov commented on src/ispc.h:
```diff
@@ -776,8 +776,12 @@ struct Globals {
+
     /** Require generation of DWARF of certain version (2, 3, 4). For
-        default version, this field is set to 0. */
+        default version, this field is set by 3. */
```
fixed
